### PR TITLE
impl(common): file-sourced subject tokens

### PR DESCRIPTION
**Pull request by coryan**
To support external accounts credentials (aka BYOID) we need to support reading the subject token from a file. The file itself can be in either text or JSON format.

Part of the work for #5915

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10277)

### CODE REVIEW

File: google/cloud/internal/external_account_token_source_file_test.cc [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+  auto const token = std::string{"a-test-only-token"};
+  std::ofstream(token_filename) << token;
+  auto const creds = nlohmann::json{{"file", token_filename}};
+  auto const source =
+      MakeExternalAccountTokenSourceFile(creds, MakeTestErrorContext());
+  ASSERT_STATUS_OK(source);
+  auto const actual = (*source)(Options{});
+  ASSERT_STATUS_OK(actual);
+  EXPECT_EQ(*actual, internal::SubjectToken{token});
+  EXPECT_EQ(std::remove(token_filename.c_str()), 0);
```
Comment (devbww) [inline google/cloud/internal/external_account_token_source_file_test.cc:56]: `#include <cstdio>`

File: google/cloud/internal/oauth2_external_account_token_source.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ * - A simple file that contains the token, maybe as a field in some JSON object
+ * - A local program may generate the token, maybe returned as a field in a JSON
+ *   object
+ *
+ * This type encapsulates all this complexity behind a simple API. The functor
+ * is created as part of parsing the external account configuration. To use it
+ * one needs to just make a call.
+ *
+ * [RFC 8663]: https://www.rfc-editor.org/rfc/rfc8693.html
+ *
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/oauth2_external_account_token_source.h]: Done

File: google/cloud/internal/oauth2_external_account_token_source.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * convert a "subject token" into an "access token". The latter is used
+ * (as one would expect) to access GCP services.
+ *
+ * External accounts may obtain the subject tokens from several different
+ * sources. The configuration for these sources can get quite involved, and the
+ * sequence of HTTP requests to get a token can be fairly involved too. These
+ * sources include:
+ * - A single HTTP request, maybe to a local metadata service or OIDC service
+ * - Multiple HTTP requests, where data obtained from previous requests is
+ *   used in the next requests (AWS works like this).
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/oauth2_external_account_token_source.h:42]: Done.

File: google/cloud/internal/external_account_token_source_file.cc [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+}  // namespace
+
+StatusOr<ExternalAccountTokenSource> MakeExternalAccountTokenSourceFile(
+    nlohmann::json const& credentials_source,
+    internal::ErrorContext const& ec) {
+  auto file =
+      ValidateStringField(credentials_source, "file", "credentials_source", ec);
+  if (!file) return std::move(file).status();
+
+  // Make a copy, most of the time this function should succeed, and we need the
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/internal/external_account_token_source_file.cc]: Done.

File: google/cloud/internal/external_account_token_source_file_test.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/internal/external_account_token_source_file_test.cc:57]: Fixed.

**Review by devbww — APPROVED** [review]

**Pull request merged**